edgedb-client/src/builder.rs

```diff
@@ -322,6 +329,171 @@ impl fmt::Display for DisplayAddr<'_> {
     }
 }
 
+impl DSN {
+    fn new(dsn: &str) -> Result<Self, Error> {
+        let admin = dsn.starts_with("edgedbadmin://");
+        if !dsn.starts_with("edgedb://") && !admin {
+            return Err(ClientError::with_message(format!(
+                "String {:?} is not a valid DSN",
+                dsn,
+            )));
+        };
+        let url = url::Url::parse(dsn).map_err(|e| {
+            ClientError::with_source(e).context(format!("cannot parse DSN {:?}", dsn))
+        })?;
+        let query = url
+            .query_pairs()
+            .into_owned()
+            .into_iter()
+            .collect::<HashMap<_, _>>();
+        Ok(Self { url, admin, query })
+    }
+
+    async fn get_value<T>(
+        &mut self,
+        key: &'static str,
+        mut v: Option<T>,
+        conv: impl FnOnce(String) -> Result<T, Error>,
+        default: impl FnOnce() -> T,
+    ) -> Result<T, Error> {
+        let v_query = self.query.remove(key);
+        let k_env = format!("{key}_env");
+        let v_env = self.query.remove(k_env.as_str());
+        let k_file = format!("{key}_file");
+        let v_file = self.query.remove(k_file.as_str());
+        let mut key_used = key;
+        let mut rv = None;
+
+        if v.is_none() {
+            key_used = key;
+            rv = v_query;
+        } else if v_query.is_some() {
+            return Err(InterfaceError::with_message(format!(
+                "redundant query in DSN: {key}"
+            )));
+        }
+        if v.is_none() && rv.is_none() {
+            if let Some(env_name) = v_env {
+                key_used = &k_env;
+                rv = get_env(&env_name)?;
+            }
+        } else if v_env.is_some() {
+            return Err(InterfaceError::with_message(format!(
+                "redundant query in DSN: {k_env}"
+            )));
+        }
+        if v.is_none() && rv.is_none() {
+            if let Some(file_path) = v_file {
+                key_used = &k_file;
+                rv = Some(
+                    fs::read_to_string(Path::new(&file_path))
+                        .await
+                        .map_err(|e| {
+                            ClientError::with_source(e)
+                                .context(format!("error reading {k_file}={file_path}"))
+                        })?,
+                );
+            }
+        } else if v_file.is_some() {
+            return Err(InterfaceError::with_message(format!(
+                "redundant query in DSN: {k_file}"
+            )));
+        }
+        if v.is_none() {
+            if let Some(rv) = rv {
+                let rv = conv(rv).with_context(|| {
+                    format!("failed to parse value of query parameter {}", key_used)
+                })?;
+                v = Some(rv);
+            }
+        }
+        Ok(v.unwrap_or_else(default))
```

Comment:
Honestly, this logic is hard to follow. And I'm not sure that this is inherent complexity.

What is the difference between `v` and `rv`? Why are you mutating both?

Logic here should be either:
```rust
let result = /* less priority value */;
if /* cond 1 */ {
   result = /* next priority */;
}
if /* cond 2 */ {
   result = /* next priority */;
}
Ok(result)
```
Or vice versa:
```result
let result = /* most priority value*/;
if result.is_none() && /* cond 1 */ {
    result = Some(...);
}
if result.is_none() && /* cond 2 */ {
    result = Some(...);
}
Ok(result.unwrap_or(/*default*/))
```
Usually the second style is used when some things are computationally intensive.

Reply:
You can also check for redundant query parameters at the start of the function and don't intermix it with the parsing logic, I think.

Reply:
Also can redundant parameters be checked in the loop at the start? I.e. any `xxx`, `xxx_env` and `xxx_file` are redundant, right? And then you can only check redundancy with the url's own value (like `host`) in this function.

But maybe moving checks to the start of `get_value` is also fine.